libgalois/src/analytics/k_truss/k_truss.cpp

```diff
@@ -0,0 +1,459 @@
+    while (srcI != srcE && (g.GetEdgeData<EdgeFlag>(srcI) & removed)) {
+      ++srcI;
+    }
+    while (dstI != dstE && (g.GetEdgeData<EdgeFlag>(dstI) & removed)) {
+      ++dstI;
+    }
+
+    if (srcI == srcE || dstI == dstE) {
+      return numValidEqual >= j;
+    }
+
+    //! Check for intersection.
+    auto sN = *g.GetEdgeDest(srcI), dN = *g.GetEdgeDest(dstI);
+    if (sN < dN) {
+      ++srcI;
+    } else if (dN < sN) {
+      ++dstI;
+    } else {
+      numValidEqual += 1;
+      if (numValidEqual >= j) {
+        return true;
+      }
+      ++srcI;
+      ++dstI;
+    }
+  }
+
+  return numValidEqual >= j;
+}
+
+/**
+ * BSPTrussJacobiAlgo:
+ * 1. Scan for unsupported edges.
+ * 2. If no unsupported edges are found, done.
+ * 3. Remove unsupported edges in a separated loop.
+ *    TODO why would it be processed in a separted loop?
+ * 4. Go back to 1.
+ */
+struct BSPTrussJacobiAlgo {
+  std::string name() { return "bsp"; }
+
+  struct PickUnsupportedEdges {
+    Graph* g;
+    unsigned int j;
+    EdgeVec& r;  ///< unsupported
+    EdgeVec& s;  ///< next
+
+    PickUnsupportedEdges(Graph* g, unsigned int j, EdgeVec& r, EdgeVec& s)
+        : g(g), j(j), r(r), s(s) {}
+
+    void operator()(Edge e) {
+      EdgeVec& w = IsSupportNoLessThanJ(*g, e.first, e.second, j) ? s : r;
+      w.push_back(e);
+    }
+  };
+
+  void operator()(Graph* g, unsigned int k) {
+    if (k - 2 == 0) {
+      return;
+    }
+
+    EdgeVec unsupported, work[2];
+    EdgeVec *cur = &work[0], *next = &work[1];
+
+    //! Symmetry breaking:
+    //! Consider only edges (i, j) where i < j.
+    katana::do_all(
+        katana::iterate(*g),
+        [&](GNode n) {
+          for (auto e : g->edges(n)) {
+            auto dest = g->GetEdgeDest(e);
+            if (*dest > n) {
+              cur->push_back(std::make_pair(n, *dest));
+            }
+          }
+        },
+        katana::steal());
+
+    while (true) {
+      katana::do_all(
+          katana::iterate(*cur),
+          PickUnsupportedEdges{g, k - 2, unsupported, *next}, katana::steal());
+
+      if (std::distance(unsupported.begin(), unsupported.end()) == 0) {
+        break;
+      }
+
+      //! Mark unsupported edges as removed.
+      katana::do_all(
+          katana::iterate(unsupported),
+          [&](Edge e) {
+            g->template GetEdgeData<EdgeFlag>(
+                katana::FindEdgeSortedByDest(*g, e.first, e.second)) = removed;
+            g->template GetEdgeData<EdgeFlag>(
+                katana::FindEdgeSortedByDest(*g, e.second, e.first)) = removed;
+          },
+          katana::steal());
+
+      unsupported.clear();
+      cur->clear();
+      std::swap(cur, next);
+    }
+  }  ///< End operator()
+};   ///< End struct BSPTrussJacobiAlgo
+
+/**
+ * BSPTrussAlgo:
+ * 1. Keep supported edges and remove unsupported edges.
+ * 2. If all edges are kept, done.
+ * 3. Go back to 3.
+ */
+struct BSPTrussAlgo {
+  std::string name() { return "bsp"; }
```

Comment:
Nit: `name()` is never used. And if that's removed then we don't need the struct and can use a real function instead.